qwerty_ast/src/typecheck.rs

```diff
@@ -0,0 +1,456 @@
+//! Qwerty typechecker implementation: walks the AST and enforces all typing rules.
+
+use crate::ast::*;
+use crate::error::{TypeError, TypeErrorKind};
+use std::collections::HashMap;
+
+//
+// ─── TYPE ENVIRONMENT ───────────────────────────────────────────────────────────
+//
+
+/// Tracks variable bindings (and potentially functions, quantum registers, etc.)
+#[derive(Debug, Clone)]
+pub struct TypeEnv {
+    vars: HashMap<String, Type>,
+    // TODO: Extend as needed (functions, modules, scopes, etc.)
+}
+
+impl TypeEnv {
+    pub fn new() -> Self {
+        Self { vars: HashMap::new() }
+    }
+
+    // Allows Shadowing
+    // (If a variable with the same name as lhs already exists in the environment, this will overwrite (shadow) the previous binding with the new type)
+    pub fn insert_var(&mut self, name: &str, typ: Type) {
```

Comment:
We already talked about this in the meeting, but we should check here if `name` is a linear type and hasn't been used yet